module/synchronization/core_rapid_test.go

```diff
@@ -0,0 +1,120 @@
+package synchronization
+
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"pgregory.net/rapid"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// TODO: remove this restriction and write a proper parametrizable generator (à la SliceOfN) once we have
+// a FlatMap combinator in rapid
+const NUM_BLOCKS int = 100
+
+// This returns a forest of blocks, some of which are in a parent relationship
+// It should include forks
+func populatedBlockStore(t *rapid.T) []flow.Header {
+	store := []flow.Header{unittest.BlockHeaderFixture()}
+	for i := 1; i < NUM_BLOCKS; i++ {
+		b := rapid.OneOf(rapid.Just(unittest.BlockHeaderFixture()), rapid.SampledFrom(store)).Draw(t, "parent").(flow.Header)
+		store = append(store, unittest.BlockHeaderWithParentFixture(&b))
+	}
+	return store
+}
+
+type rapidSync struct {
+	store          []flow.Header
+	core           *Core
+	idRequests     map[flow.Identifier]int // pushdown automaton to track ID requests
+	heightRequests map[uint64]int          // pushdown automaton to track height requests
+}
+
+// Init is an action for initializing a rapidSync instance.
+func (r *rapidSync) Init(t *rapid.T) {
+	var err error
+
+	r.core, err = New(zerolog.New(ioutil.Discard), DefaultConfig())
+	require.NoError(t, err)
+
+	r.store = populatedBlockStore(t)
+	r.idRequests = make(map[flow.Identifier]int)
+	r.heightRequests = make(map[uint64]int)
+}
+
+// RequestByID is an action that requests a block by its ID.
+func (r *rapidSync) RequestByID(t *rapid.T) {
+	b := rapid.SampledFrom(r.store).Draw(t, "id_request").(flow.Header)
+	r.core.RequestBlock(b.ID())
+	// Re-queueing by ID should always succeed
+	r.idRequests[b.ID()] = 1
+	// Re-qeueuing by ID "forgets" a past height request
+	r.heightRequests[b.Height] = 0
+}
+
+// RequestByHeight is an action that requests a specific height
+func (r *rapidSync) RequestByHeight(t *rapid.T) {
+	b := rapid.SampledFrom(r.store).Draw(t, "id_request").(flow.Header)
```

Comment:
What is this `Id_request` for? Does it mean `Draw` with the same value will take the same sample?